src/net_processing.cpp

```diff
@@ -5246,6 +5247,28 @@ void PeerManagerImpl::EvictExtraOutboundPeers(std::chrono::seconds now)
             }
         }
     }
+
+    // If we are at the max full outbound limit but not above, check if any peers
+    // don't support tx relay (they might be in -blocksonly mode) - if so,
+    // evict the newest of them so we can fill the slot with a tx-relaying outbound peer
+    // This is not done if we are in -blocksonly mode ourselves or still in IBD,
+    // because then we don't care if our peer supports tx relay.
+    if (extra_outbound_count == 0 && !m_opts.ignore_incoming_txs && !m_chainman.IsInitialBlockDownload()) {
+        std::optional<NodeId> node_to_evict;
+        m_connman.ForEachNode([&node_to_evict](CNode* node) {
+            if (!node->IsFullOutboundConn() || node->m_relays_txs) return;
+            if (!node_to_evict.has_value() || *node_to_evict < node->GetId()) {
+                node_to_evict = node->GetId();
+            }
+        });
```

Comment:
What if there is, at least, a node already flagged to be disconnected? Shouldn't we skip this altogether? Otherwise, it could be the case that we end up evicting more nodes that we really need

Reply:
Nvm, `ForEachNode` already checks `NodeFullyConnected`

Reply:
Yes, that's why I didn't include the check here (the similar checks a few lines above aren't necessary either)

Reply:
Assume `fDisconnect` could be useful here too. Just in case `ForEachNode` gets updated.

Reply:
I also don't see that as a real risk, but I do see the problem is that whoever reads this code (or decides to use `ForEachNode` somewhere else) could easily get confused, especially since some callers check again for `fDisconnect`.
I added a refactor commit renaming `ForEachNode()` to `ForEachFullyConnectedNode()` and removing the duplicate checks / comments. What do you think about that approach?

Reply:
To be clear, i suggested literally having `Assume(fDisconnect)`, not an extra `if` check (that would actually make code worse).

Even if you refactor it into `ForEachFullyConnectedNode`, I think this `Assume(fDisconnect)` is kinda useful — because who knows what exactly `ForEachFullyConnectedNode` means :)

But I'm fine either way, either you rename or add `Assume` (both are ideal i think :))